### DISCUSSION TITLE
Does SeaweedFS S3 Support Conditional PutObject Operations?

### DISCUSSION THREAD

**Discussion by dpeckett** [seaweedfs/seaweedfs#5299] [General]
It might be a little mad scientist but I've been looking at implementing a locking library on top of S3. Amazon stubbornly refuses to provide atomic operations for S3 (because they know it'd be the end of their eventual consistency empire) but GCS and Azure have no qualms providing them, and most third-party S3 providers seem to support the atomic PutObject extensions from Swift.

Is this implemented in SeaweedFS?

Eg. https://developers.cloudflare.com/r2/api/s3/extensions/#putobject-1

**Comment by legout** [comment]
I am also interested in this. Any updates?

**Reply by kmlebedev** [reply]
To be honest, I don't understand what exactly you're talking about?
Is there a specification for such a request, how to reproduce it?
Is there a corresponding test in https://github.com/ceph/s3-tests?

**Reply by legout** [reply]
I am sorry. Let me give you some more context.

delta-rs guarantee ACID transactions on s3, as explained in the delta-rs documentation:

_Delta lake guarantees [ACID transactions](https://delta-io.github.io/delta-rs/how-delta-lake-works/delta-lake-acid-transactions/) when writing data. This is done by default when writing to all supported object stores except AWS S3. (Some S3 clients like **CloudFlare R2** or **MinIO** may enable concurrent writing without a locking provider, refer to [this section](https://delta-io.github.io/delta-rs/usage/writing/writing-to-s3-with-locking-provider/#enabling-concurrent-writes-for-alternative-clients) for more information)._

Therefore, one has to set the following in `storage_options` when using R2 or Minio.
```python
storage_options = {
    "conditional_put": "etag",
}
```

This feature is described in the **R2** documentaion here: [Conditional operations in PutObject](https://developers.cloudflare.com/r2/api/s3/extensions/#conditional-operations-in-putobject)

That is why I am interested, whether SeaweedFS also supports this feature or not.

**Reply by chrislusf** [reply]
there is already function checkPreconditions for GET or HEAD https://github.com/seaweedfs/seaweedfs/blob/5bd5a76d252421256f8f0d71188985efe1e7b7da/weed/server/filer_server_handlers_read.go#L38

need to add it to PUT and need some test cases to confirm the change.

**Reply by legout** [reply]
Awesome!

**Reply by legout** [reply]
@chrislusf Can you already say, when this will be available in seaweedfs? Should I open a seperate issue?

**Comment by Material-Scientist** [comment]
You could use something like fsspec's [semi-atomic transactions](https://filesystem-spec.readthedocs.io/en/latest/features.html#transactions).

Basically puts your data in a temp dir until you're ready to write it to the actual dir.

**Comment by OrkhanAlikhanov** [comment]
>Amazon stubbornly refuses to provide atomic operations for S3

@dpeckett They seem to support it now: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#AmazonS3-PutObject-request-header-IfNoneMatch

I tried it with seaweedfs, does not seem to support it with PUT operation. Also tried with IBM COS, it does support it.

**Comment by kmlebedev** [comment]
@OrkhanAlikhanov Hi, please write in outline how you tested

It seems that the issue is with returning the ConditionalRequestConflict error.
